functional-samples/sample.co2meter/modules/co2_meter.js

```diff
@@ -0,0 +1,164 @@
+ */
+
+import { PERMISSION_GRANTED_MESSAGE } from './constant.js';
+
+const key = new Uint8Array([0xc4, 0xc6, 0xc0, 0x92, 0x40, 0x23, 0xdc, 0x96]);
+
+function KelvinToFahrenheit(k) {
+  return Math.trunc(((k - 273.15) * 9) / 5 + 32);
+}
+
+class CO2Meter {
+  constructor() {
+    this.device = null;
+    this.connectClientCB = null;
+    this.disconnectClientCB = null;
+    this.co2ReadingClientCB = null;
+    this.tempReadingClientCB = null;
+    this.connectHandler = this.connectHandler.bind(this);
+    this.disconnectHandler = this.disconnectHandler.bind(this);
+    this.onInputReport = this.onInputReport.bind(this);
+  }
+
+  /**
+   * @description This function initializes the CO2Meter object.
+   */
+  async init(
+    connectCallback = null,
+    disconnectCallback = null,
+    co2ReadingCallback = null,
+    tempReadingCallback = null
+  ) {
+    this.connectClientCB = connectCallback;
+    this.disconnectClientCB = disconnectCallback;
+    this.co2ReadingClientCB = co2ReadingCallback;
+    this.tempReadingClientCB = tempReadingCallback;
+    navigator.hid.addEventListener('connect', this.connectHandler);
+    navigator.hid.addEventListener('disconnect', this.disconnectHandler);
+    console.log('CO2Meter init() done');
+  }
+
+  async startReading() {
+    if (this.device) {
+      console.log('CO2 reading has already started!');
+      return;
+    }
+    const devices = await navigator.hid.getDevices();
+    if (devices.length == 0) {
+      throw 'No CO2 meter for reading!';
+    }
+    this.device = devices[0];
+
+    try {
+      await this.device.open();
+      await this.device.sendFeatureReport(0, key);
+    } catch (e) {
+      console.log('CO2 reading exception:', e);
+      await this.device.close();
+      this.device = null;
+      throw 'Fail to open CO2 meter for reading!';
+    }
+
+    this.device.addEventListener('inputreport', this.onInputReport);
+  }
+
+  async stopReading() {
+    if (this.device) {
+      this.device.removeEventListener('inputreport', this.onInputReport);
+      await this.device.close();
+      this.device = null;
+    }
+  }
+
+  onInputReport(report) {
+    let data = new Uint8Array(
+      report.data.buffer,
+      report.data.byteOffset,
+      report.data.byteLength
+    );
+
+    const op = data[0];
+    let val = (data[1] << 8) | data[2];
+
+    if (op == 0x50) {
+      console.log(`Current CO2 reading is ${val}`);
+      if (this.co2ReadingClientCB) {
+        this.co2ReadingClientCB(val);
+      }
+    } else if (op == 0x42) {
+      val = val / 16;
+      console.log(`Current Temp reading is ${val}`);
+      if (this.tempReadingClientCB) {
+        this.tempReadingClientCB(val);
+      }
+    }
+  }
+
+  /**
+   * @description Request user to grant permission for using CO2 meter.
+   * The extension currently only support this model:
+   * https://www.co2meter.com/products/co2mini-co2-indoor-air-quality-monitor
+   */
+  requestPermission() {
+    navigator.hid
```

Comment:
Could we use async/await here for consistency?

Reply:
Done.